### PR TITLE
DAOS-4460 csum: Checksum Calculations on Fetch Account for Holes

### PR DESCRIPTION
**Pull request by ryon-jensen**
Use the IO Map on a fetch to calculate checksums for what was
actually returned, ignoring holes.

- Added an io map parameter to appropriate csum methods
- Client requests map on fetch if checksums is enabled
- Several tests for different types of holes in different places
  wrt chunks

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

File: src/common/checksum.c [hunk 1]
```diff
@@ -588,51 +588,168 @@ daos_csummer_alloc_iods_csums(struct daos_csummer *obj, daos_iod_t *iods,
...
-	       size_t rec_len, daos_recx_t *recxs, size_t nr,
-	       struct dcs_csum_info *csums)
+calc_csum_recx_with_no_map(struct daos_csummer *obj, size_t csum_nr,
+			   daos_recx_t *recx,
+			   struct dcs_csum_info *csum_info,
+			   size_t rec_len, d_sg_list_t *sgl,
+			   uint32_t rec_chunksize,
+			   struct daos_sgl_idx *idx)
 {
+	uint32_t		 i;
```
Comment (jolivier23) [inline src/common/checksum.c:598]: personal nit...order by decreasing size, similar to what you would do for a struct

File: src/common/checksum.c [hunk 2]
```diff
@@ -588,51 +588,168 @@ daos_csummer_alloc_iods_csums(struct daos_csummer *obj, daos_iod_t *iods,
...
+						* rec_len;
+				daos_sgl_get_bytes(sgl, idx,
+						   bytes_to_skip,
+						   NULL, NULL);
+			}
+			bytes_for_csum = mapped_chunk.dcr_nr * rec_len;
+			rc = daos_sgl_processor(sgl, idx, bytes_for_csum,
+						checksum_sgl_cb, obj);
+			if (rc != 0) {
+				D_ERROR("daos_sgl_processor error: %d\n", rc);
```
Comment (jolivier23) [inline src/common/checksum.c:710]: In looking at the code, it appears nobody using a cf_finish callback....if one did, would you still want to call it on an error?

Comment (ryon-jensen) [reply src/common/checksum.c:710]: Yes, good catch. Currently nothing is using cf_finish, but the sha checksum functions will. I'll make sure to fix this when I work on that.

**Comment by ryon-jensen** [comment]
IOR/MDTest/FIO Tests fail due to DAOS-4850
Punch failure due to DAOS-4853 (Failed to stop agents)
Degraded failure not sure yet

**Pull request merged**